nifi-nar-bundles/nifi-oauth-bundle/nifi-oauth/src/main/java/org/apache/nifi/oauth/AbstractOAuthControllerService.java

```diff
@@ -0,0 +1,172 @@
+                    "that tells when the access token received will be invalidated. Comparing the current system clock to that value and understanding if the " +
+                    "access token is still valid can be achieved this way. However since the clocks can be out of sync with the authentication server " +
+                    "this property ensures that the access token can be refreshed at a configurable interval before it actual expires to ensure no downtime " +
+                    "waiting on new tokens from the authentication server.")
+            .defaultValue("10")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    protected void onEnabled(final ConfigurationContext context)
+            throws InitializationException {
+        Map<PropertyDescriptor, String> allProperties = context.getProperties();
+        Iterator<PropertyDescriptor> itr = allProperties.keySet().iterator();
+        while (itr.hasNext()) {
+            PropertyDescriptor pd = itr.next();
+            if (pd.isDynamic()) {
+                extraHeaders.put(pd.getName(), allProperties.get(pd));
+            }
+        }
+
+        this.authUrl = context.getProperty(AUTH_SERVER_URL).getValue();
+        this.expireTimeSafetyNetSeconds = new Long(context.getProperty(EXPIRE_TIME_SAFETY_NET).getValue());
+
+        // Name of the OAuth2 spec fields that should be extracted from the JSON authentication response. While in theory every provider should be
+        // using the same names in the response JSON that rarely happens. These values are used to ensure that if the provider does NOT follow the
+        // spec the user can still use this integration approach instead of having to write something custom.
+        this.accessTokenRespName = context.getProperty(RESPONSE_ACCESS_TOKEN_FIELD_NAME).getValue();
+        this.expireTimeRespName = context.getProperty(RESPONSE_EXPIRE_TIME_FIELD_NAME).getValue();
+        this.expireInRespName = context.getProperty(RESPONSE_EXPIRE_IN_FIELD_NAME).getValue();
+        this.tokenTypeRespName = context.getProperty(RESPONSE_TOKEN_TYPE_FIELD_NAME).getValue();
+        this.scopeRespName = context.getProperty(RESPONSE_SCOPE_FIELD_NAME).getValue();
+    }
+
+    public boolean isOAuthTokenExpired() {
+        if (expiresTime == -1 && expiresIn == -1)
+            return true;
+
+        if (expiresTime > 0) {
+            // Use the actual time that the token will authenticate
+            if ((expiresTime - (expireTimeSafetyNetSeconds * 1000)) <= System.currentTimeMillis()) {
+                return true;
+            } else {
+                return false;
+            }
+        } else if (expiresIn > 0) {
+            long tombStoneTS = ((lastResponseTimestamp + (expiresIn * 1000)) - (expireTimeSafetyNetSeconds * 1000));
```

Comment:
I think this logic also needs to be evaluated for bidirectional time jitter as discussed above.

Reply:
Same as above